### PR TITLE
Introduce document and tweak the design of serialization attributes

### PR DESCRIPTION
**Pull request by ArcturusZhang**
Fixes https://github.com/Azure/autorest.csharp/issues/3429

# Description

1. This PR introduces documents for the new serialization related codegen attributes.
2. This PR tweaks the design of the serialization hooks attribute, now we have to use named argument to assign its properties. Also rename the previous `SerializationHook` and `DeserializationHook` to `SerializationValueHook` and `DeserializationValueHook` since they are just hook methods for serializing/deserializing the value of the property (writing property name not included).
3. <del>This PR adds a new property `SerializationHook` to `CodeGenMemberSerializationHooks` attribute, this hook will completely replace the serialization of this property.</del> we could introduce this back later if someone asks for this.
4. This PR fixes a compilation issue when you add a nullable property into the serialization.

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -541,7 +541,523 @@ namespace Azure.Service.Models
 
 ### Customize serialization/deserialization methods
 
-Use the [Replace any generated member](#replace-any-generated-member) approach to replace Serialize/Deserialize method with a custom implementation.
+If you want to change the property name that serializes into the JSON or deserializes from the JSON, you could define your own property name using the `CodeGenMemberSerialization` attribute.
```
Comment (ArcturusZhang) [inline readme.md:544]: add a section
change property serialization/deserialization name

File: readme.md [hunk 2]
```diff
@@ -541,7 +541,523 @@ namespace Azure.Service.Models
...
+    public partial class Cat
+    {
+        /* omit the ctors for brevity */
+        public string Name { get; set; }
+        public string Color { get; set; }
+    }
+}
+```
+
+Now we want to change the property name `Name` to `catName` in the serialized JSON, we could do the following: Define a partial class and the property you want to change in the partial class, and add `CodeGenMemberSerialization` attribute to it:
```
Comment (ArcturusZhang) [inline readme.md:564]: **Add customized code** (see reference for others)

File: readme.md [hunk 3]
```diff
@@ -541,7 +541,523 @@ namespace Azure.Service.Models
...
+                    color = property.Value.GetString();
+                    continue;
+                }
+            }
+            return new Cat(name, color);
+        }
+    }
+}
+```
+
```
Comment (ArcturusZhang) [inline readme.md:636]: another section

File: readme.md [hunk 4]
```diff
@@ -541,80 +541,696 @@ namespace Azure.Service.Models
...
+            color = "black";
+            return new Cat(name, color);
+        }
+    }
+}
+```
+
+**Generated code after:**
 
 Generated code won't contain the DeserializeCat method and the custom one would be used for deserialization.
```
Comment (chunyu3) [inline readme.md:1235]: In the customized code, there is `void global:Azure.Core.IUtf8JsonSerializable.Write(Utf8JsonWriter writer)`, so the serialized code will not be generated also. Right?

Comment (ArcturusZhang) [reply readme.md:1235]: Yes, both serialization and deserialization are overridden by customized code in this case.

**Comment by ArcturusZhang** [comment]
@m-nash could you have a review on this PR?

**Review by m-nash — APPROVED** [review]

**Review by chunyu3 — APPROVED** [review]

File: readme.md (hunk 4 above)
Comment (ArcturusZhang) [reply readme.md]: I tweaked this line to emphasize both `Write` and `DeserializeCat` are gone.

**Pull request merged**